rllib/algorithms/ppo/utils/env_runner.py

```diff
@@ -0,0 +1,471 @@
+        # to be weight-synched each iteration).
+        else:
+            policy_dict, _ = self.config.get_multi_agent_setup(env=self.env)
+            module_spec = self.config.get_marl_module_spec(policy_dict=policy_dict)
+            # TODO (simon): This here is only for single agent.
+            # This is a MARL.
+            self.module: MultiAgentRLModule = module_spec.build()  # [DEFAULT_POLICY_ID]
+
+        # Let us set this as default for PPO.
+        self._needs_initial_reset: bool = True
+        self._episodes: List[Optional[Episode]] = [None for _ in range(self.num_envs)]
+
+        self._done_episodes_for_metrics: List[Episode] = []
+        self._ongoing_episodes_for_metrics: Dict[List] = defaultdict(list)
+        self._ts_since_last_metrics: int = 0
+
+    @override(EnvRunner)
+    def sample(
+        self,
+        *,
+        num_timesteps: int = None,
+        num_episodes: int = None,
+        explore: bool = True,
+        random_actions: bool = False,
+        with_render_data: bool = False,
+    ) -> Tuple[List[Episode], List[Episode]]:
+        """Runs and returns a sample (n timesteps or m episodes) on the env(s)."""
+
+        # If not execution details are provided, use self.config.
+        if num_timesteps is None and num_episodes is None:
+            if self.config.batch_mode == "truncate_episodes":
+                num_timesteps = (
+                    self.config.get_rollout_fragment_length(
+                        worker_index=self.worker_index
+                    )
+                    * self.num_envs
+                )
+            else:
+                num_episodes = self.num_envs
+
+        # Sample n timesteps.
+        if num_timesteps is not None:
+            return self._sample_timesteps(
+                num_timesteps=num_timesteps,
+                explore=explore,
+                random_actions=random_actions,
+                force_reset=False,
+            )
+        # Sample m episodes.
+        else:
+            # `_sample_episodes` returns only a single list (with completed episodes)
+            # therefore add the empty list for the truncated episodes.
+            return (
+                self._sample_episodes(
+                    num_episodes=num_episodes,
+                    explore=explore,
+                    random_actions=random_actions,
+                    with_render_data=with_render_data,
+                ),
+                [],
+            )
+
+    def _sample_timesteps(
+        self,
+        num_timesteps: int,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+    ) -> Tuple[List[Episode], List[Episode]]:
+        """Helper method to sample n timesteps."""
+
+        done_episodes_to_return = []
+
+        # Get initial states for all 'batch_size_B` rows in the forward batch,
+        # i.e. for all vector sub_envs.
+        initial_states = tree.map_structure(
```

Comment:
Thought: Let's just do this iff the marl_module is recurrent. Otherwise, we should set initial_states/states everywhere to None.

Reply:
Or maybe better, use an empty dict iff not recurrent. This would save maybe a lot of if/else stuff below b/c then it won't affect the `tree.map_structure` calls everywhere.